Complete regex literal lexing

hamishknight opened the pull request:
The previous regex literal lexing logic didn't implement all the heuristics needed to parse all the regex literals that the old C++ parser handles. Update it with the heuristics it needs, and better handle diagnostics for invalid cases.

This unfortunately requires a bit of lexical hackery, including a heuristic to classify previous token kinds to determine if we're in regex literal position, but it's needed to handle pathological cases such as `x /^ y/` where there's an ambiguity with a binary operator.

Resolves #1059
rdar://102213809

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test Windows

hamishknight commented:
@swift-ci please test Windows

hamishknight commented on CodeGeneration/Sources/SyntaxSupport/TokenSpec.swift:
```diff
@@ -178,6 +178,7 @@ public let SYNTAX_TOKENS: [TokenSpec] = [
   PunctuatorSpec(name: "Ellipsis", kind: "ellipsis", text: "..."),
   PunctuatorSpec(name: "Equal", kind: "equal", text: "=", requiresLeadingSpace: true, requiresTrailingSpace: true),
   PunctuatorSpec(name: "ExclamationMark", kind: "exclaim_postfix", text: "!"),
+  MiscSpec(name: "ExtendedRegexDelimiter", kind: "extended_regex_delimiter", nameForDiagnostics: "extended delimiter", classification: "RegexLiteral"),
```
This is intentional, for strings the presence of `#` changes how escape sequences work, for regexes it has no effect on how something like `\n` works. It's just purely a way of changing the delimiter so you can write `/` in the regex without escaping.

hamishknight replied:
I should also note it's the terminology we used for the evolution proposal https://github.com/apple/swift-evolution/blob/main/proposals/0354-regex-literals.md#extended-delimiters--

ahoppen replied:
OK, just wanted to double-check.

hamishknight commented on Sources/SwiftParser/Expressions.swift:
```diff
@@ -1433,13 +1433,32 @@ extension Parser {
+    let openPounds = self.consume(if: .extendedRegexDelimiter)
+
+    // Parse the opening slash.
+    let (unexpectedBeforeSlash, openSlash) = self.expect(.regexSlash)
```
Yeah, though the lexer should ensure we never produce a `extendedRegexDelimiter` without a following `regexSlash`. I'm happy to add a check for it tho

hamishknight commented on Sources/SwiftParser/Lexer/Cursor.swift:
```diff
@@ -217,6 +239,8 @@ extension Lexer {
 
     /// If we have already lexed a token, the kind of the previously lexed token
     var previousTokenKind: RawTokenKind?
+    var previousKeyword: Keyword?
```
Yes, it's `nil` in that case. I can add a comment

hamishknight commented on Sources/SwiftParser/Lexer/Cursor.swift:
```diff
@@ -776,7 +822,10 @@ extension Lexer.Cursor {
-  private mutating func lexNormal(sourceBufferStart: Lexer.Cursor) -> Lexer.Result {
+  private mutating func lexNormal(
+    sourceBufferStart: Lexer.Cursor,
+    preferRegexOverBinaryOperator: Bool = false
```
I'm fine changing it, but the only caller should ever be the `case .preferRegexOverBinaryOperator` in `nextToken`, and we don't to proliferate this hack anywhere else

ahoppen replied:
That’s also true. I’m just a little worried we’ll refactor something and then forget to pass this parameter down the chain.

hamishknight commented on Sources/SwiftParser/Lexer/LexemeSequence.swift:
```diff
@@ -29,7 +29,7 @@ extension Lexer {
     /// The memory footpring of not freeing past lexer states is neglible. It's
     /// usually less than 0.1% of the memory allocated by the syntax arena.
-    var lexerStateAllocator = BumpPtrAllocator(slabSize: 256)
+    var lexerStateAllocator = BumpPtrAllocator(slabSize: 512)
```
Mainly a gut feeling, I did benchmark, and there was a slight improvement, but it was pretty close to margin of error tbh. I'm okay not changing this. `RegexLiteralLexemes` is 7 words, so this accommodates 8 regex literals before needing another slab (including two words for the state itself).

ahoppen replied:
I’m happy with either value. I was just interested in what made you change it. I'll trust your judgment. Just remember that most code doesn’t contain regex literals, so we shouldn’t be optimizing too much for regex literals being present.

hamishknight replied:
That's fair, though this is also bearing in mind that we're now also burning 2 words every time we see `try?`/`try!`. So even without regex literals, a slab size of 256 only accommodates 16 string literals/`try?`/`try!`. That's probably fine, but 32 feels a bit better. I really don't feel too strongly about this though.

ahoppen replied:
I don’t think we need to allocate 2 words for every `try?`/`try!` because we can store a single state inline with the `StateStack` without having to do any allocation on the bump allocator.

ahoppen replied:
The same goes for the regex state since you don’t nest regex states. So I wouldn’t expect us to actually hit the bump allocator for regexes at all.

hamishknight replied:
> I don’t think we need to allocate 2 words for every `try?`/`try!` because we can store a single state inline with the StateStack without having to do any allocation on the bump allocator.

Hmm I guess we could do that if we know we're already in the normal lexing state, but we wouldn't be able to do that if we're in an interpolation, right?

> The same goes for the regex state since you don’t nest regex states. So I wouldn’t expect us to actually hit the bump allocator for regexes at all.

Same as above, also we'd still need to allocate the lexemes struct on the allocator as we can't store that inline.

hamishknight replied:
Ahhhh I realize I misunderstood the state transition code, pushing a state only allocates when there are n > 1 states, so it doesn't affect `try?`/`try!`/regex literals. We still need to allocate the lexemes themselves on the bump pointer allocator, but I think this change probably is less compelling with that being the only thing to consider. I'll revert it.

ahoppen replied:
I was just writing a comment saying exactly that.

hamishknight commented on Sources/SwiftParser/Lexer/Cursor.swift:
```diff
@@ -1919,7 +1957,37 @@ extension Lexer.Cursor {
+      case UInt8(ascii: "?"):
+        return .postfixQuestionMark
+      default:
+        fatalError("Must be at '!' or '?'")
```
Personally `fatalError` feels more "this shouldn't be reachable" whereas `preconditionFailure` feels more "you violated the contract of an API". This definitely fits into the former, but I'm happy to change if you want.

ahoppen replied:
I just think we should use one or the other because otherwise I just always start wondering: Why did someone choose `fatalError` here instead of `preconditionFailure`, which is most of the time completely pointless to wonder about.

hamishknight commented on Sources/SwiftParser/Lexer/Cursor.swift:
```diff
@@ -1934,15 +2002,31 @@ extension Lexer.Cursor {
+    do {
+      var ptr = tokStart
+      while ptr.input.baseAddress! < self.input.baseAddress! {
+        // Scan for the first '/' in the operator to see if it starts a regex
+        // literal.
```
Hmm not sure I understand, this logic handles things like `^^/.../` where we need to scan to split the `/.../` from a prefix operator `^^`. This code path doesn't handle extended literals, we can always lex those directly as `#` isn't an operator character.

ahoppen replied:
Ah, I see. Makes sense. I didn’t think about splitting the regex from the operator here.

hamishknight commented on Tests/SwiftParserTest/LexerTests.swift:
```diff
@@ -422,6 +456,117 @@ public class LexerTests: XCTestCase {
+    assertLexemes(
+      "^/*/",
+      lexemes: [
+        LexemeSpec(.binaryOperator, text: "^", trailing: "/*/")
```
Oh heh, good catch! I was thinking it was just unterminated, but yeah we don't record an error or anything. Oops.

hamishknight commented on Tests/SwiftParserTest/translated/ForwardSlashRegexSkippingAllowedTests.swift:
```diff
@@ -0,0 +1,157 @@
+import XCTest
+
+final class ForwardSlashRegexSkippingAllowedTests: XCTestCase {
+  func testForwardSlashRegexSkippingAllowed1() {
+    assertParse(
+      """
+      // Make sure we can skip in all of the below cases.
+      """
+    )
+  }
+
+  func testForwardSlashRegexSkippingAllowed2() {
+    assertParse(
+      #"""
+      // The printing implementation differs in asserts and no-asserts builds, it will
+      // either print `"Parse.NumFunctionsParsed" 0` or not print it at all. Make sure
+      // we don't output any non-zero value.
+      // CHECK-NOT: {{"Parse.NumFunctionsParsed" [^0]}}
+      """#
+    )
+  }
+
```
Good catch, I thought I got rid of most of the pointless ones, but I guess I missed a few

hamishknight commented on Tests/SwiftParserTest/translated/ForwardSlashRegexSkippingInvalidTests.swift:
```diff
@@ -0,0 +1,280 @@
+      func d() { _ = /2️⃣ x{*/ }
+      """,
+      diagnostics: [
+        DiagnosticSpec(locationMarker: "1️⃣", message: "bare slash regex literal may not start with space"),
```
Not yet, I'm leaving that as future work for now, there's a `TODO` comment where we produce the diagnostic, I'll file an issue.

hamishknight commented on Sources/SwiftParser/Lexer/RegexLiteralLexer.swift:
```diff
@@ -0,0 +1,786 @@
+  }
+
+  /// Attempt to lex a character of the regex pattern.
+  mutating func lexPatternCharacter(escaped: Bool = false) -> LexResult {
```
Sure, I mean it's not that big of a deal as `RegexLiteralLexer` is `fileprivate`, but I guess there's no harm

ahoppen replied:
Oh, I didn’t realize `RegexLiteralLexer` was `fileprivate`. I assumed it was internal. Ignore what I said.

hamishknight commented on Sources/SwiftParser/Lexer/RegexLiteralLexer.swift:
```diff
@@ -0,0 +1,786 @@
+  /// Tracks the current group depth, used to enforce the heuristic that a bare
+  /// slash regex literal with an unbalanced ')' should be treated as an
+  /// operator instead.
+  private var groupDepth = 0
+  private var customCharacterClassDepth = 0
```
Sure

hamishknight commented on Sources/SwiftParser/Lexer/RegexLiteralLexer.swift:
```diff
@@ -0,0 +1,786 @@
+      groupDepth -= 1
+
+    case "[" where !escaped:
+      customCharacterClassDepth += 1
```
Yes! You can even do set operations `/[[ab]--[bc]]/ == /[a]/`

ahoppen replied:
Crazy 🤯

hamishknight commented on Sources/SwiftParser/Lexer/RegexLiteralLexer.swift:
```diff
@@ -0,0 +1,786 @@
+  /// If this is a multi-line regex literal, the cursor position of the first
+  /// newline.
+  private var firstNewline: Lexer.Cursor?
+  private var isMultiline: Bool { firstNewline != nil }
```
We allow it, but the closing delimiter is `/#`, not `#/`. We don't check indentation, because we ignore all whitespace in multi-line regex.

ahoppen replied:
Stupid me for messing up the closing delimiter. And interesting that we don’t enforce indentation. But it makes parsing multi-line regex literals a lot simpler 🤩

hamishknight commented on Sources/SwiftParser/Lexer/RegexLiteralLexer.swift:
```diff
@@ -0,0 +1,786 @@
+          // }
+          //
+          if mustBeRegex {
+            // TODO: We ought to have a fix-it that inserts a backslash to escape.
```
Yes, I was meaning to, I think the `TODO`s are still useful for people reading through the code tho, I wouldn't want to get rid of them.

hamishknight commented on Sources/SwiftParser/Lexer/RegexLiteralLexer.swift:
```diff
@@ -0,0 +1,786 @@
+          _ = newlineScan.advance()
+          continue
+        case "\n", "\r":
+          firstNewline = newlineScan
```
Hmm it won't even consume `\r`, right? It should place the cursor at `\r`, and they should then be lexed by `lexPatternCharacter`.

ahoppen replied:
Ah, yes, you’re right. Thanks for the explanation. I just had a few issues where string literals were not handling `\r\n` properly so this jumped at me.

hamishknight commented on Sources/SwiftParser/Lexer/RegexLiteralLexer.swift:
```diff
@@ -0,0 +1,786 @@
+  }
+
+  mutating func recordCloseSlash(endingAt closeSlashEnd: Lexer.Cursor.Position) {
+    precondition(lastLexemeKind == .openingSlash)
```
No, we use the recording of the close slash to push the pattern token (as we need the close slash to know the length). A previous version of this pushed an "incomplete" pattern token, but it was more complicated doing it that way.

ahoppen replied:
Ah, I see. Makes sense 👍🏽

hamishknight commented on Sources/SwiftParser/Lexer/Cursor.swift:
```diff
@@ -1934,15 +2002,31 @@ extension Lexer.Cursor {
+          _ = ptr.advance()
+          continue
+        }
+        guard
+          let result = self.tryLexOperatorAsRegexLiteral(
+            at: ptr,
+            operatorStart: tokStart,
+            operatorEnd: self,
+            sourceBufferStart: sourceBufferStart,
+            preferRegexOverBinaryOperator: preferRegexOverBinaryOperator
+          )
+        else { break }
+        return result
```
Hmm honestly I wouldn't say it's cleaner, we'll have to check for `regexScan.input.baseAddress! < self.input.baseAddress!` twice. We could add a utility to `Lexer.Cursor` for this tho if you want.

ahoppen commented on Sources/SwiftParser/Lexer/RegexLiteralLexer.swift:
```diff
@@ -0,0 +1,786 @@
+    at cursor: Lexer.Cursor
+  ) {
+    precondition(lastLexemeKind == .openingSlash)
+    patternError = .init(kind, position: cursor)
```
Oh, do they also prefer the last error. That might explain a few weird errors I was seeing in the tests. I think we should always prefer the first, both for regex and for string literals.

hamishknight replied:
Filed #1480 for string literals

hamishknight commented on Tests/SwiftParserTest/translated/ForwardSlashRegexSkippingTests.swift:
```diff
@@ -0,0 +1,462 @@
+      func a4() { _ = /1️⃣ / }
+      """,
+      diagnostics: [
+        // TODO: Old parser had a fix-it to add backslash to escape
```
Filed #1479

hamishknight commented:
https://github.com/apple/sourcekit-lsp/pull/724

@swift-ci please test

hamishknight commented:
@ahoppen Addressed review feedback, let me know what you think. Also noticed that we needed to properly classify the prefix operators we're splitting, so pushed a commit to fix that.

hamishknight commented:
https://github.com/apple/sourcekit-lsp/pull/724

@swift-ci please test Windows

ahoppen reviewed (commented):
I kept two discussions as unresolved (bump pointer allocator slab size + performance impact of passing `previousKeyword` around). Once those are resolved, this looks good to me.

hamishknight commented on Sources/SwiftParser/Lexer/Cursor.swift:
```diff
@@ -385,6 +419,10 @@ extension Lexer.Cursor {
       cursor: cursor
     )
+    self.previousTokenKind = result.tokenKind
+    self.previousKeyword = result.tokenKind == .keyword ? Keyword(lexeme.tokenText)! : nil
```
Sorry, yes, it is a slight improvement, from ~18.5ms to ~18.2ms. I think it's worth keeping the change.

ahoppen replied:
OK, that’s in line with what I would have expected. Thanks for verifying.

hamishknight commented:
https://github.com/apple/sourcekit-lsp/pull/724

@swift-ci please test

hamishknight commented:
https://github.com/apple/sourcekit-lsp/pull/724

@swift-ci please test Windows

hamishknight commented:
https://github.com/apple/sourcekit-lsp/pull/724

@swift-ci please test Windows

hamishknight commented:
https://github.com/apple/sourcekit-lsp/pull/724

@swift-ci please test Windows

hamishknight commented:
https://github.com/apple/sourcekit-lsp/pull/724

@swift-ci please test Windows

hamishknight commented:
https://github.com/apple/sourcekit-lsp/pull/724

@swift-ci please test Windows

hamishknight commented:
https://github.com/apple/sourcekit-lsp/pull/724

@swift-ci please test Windows

hamishknight commented:
https://github.com/apple/sourcekit-lsp/pull/724

@swift-ci please test Windows
